core/src/main/java/org/infinispan/interceptors/impl/CacheMgmtInterceptor.java

```diff
@@ -674,10 +684,81 @@ public long getAverageRemoveTimeNanos() {
       return counters.get(StripeB.removeTimesFieldUpdater) / removes;
    }
 
+   @ManagedAttribute(
+         description = "Approximate number of entries currently in the cache, including persisted and expired entries",
+         displayName = "Approximate number of entries"
+   )
+   public long getApproximateEntries() {
+      // Don't restrict the segments in case some writes used CACHE_MODE_LOCAL
+      IntSet allSegments = IntSets.immutableRangeSet(cacheConfiguration.clustering().hash().numSegments());
+      // Do restrict the segments when counting entries in shared stores
+      IntSet writeSegments;
+      if (distributionManager != null) {
+         writeSegments = distributionManager.getCacheTopology().getLocalWriteSegments();
+      } else {
+         writeSegments = allSegments;
+      }
+      long persistenceSize = CompletionStages.join(approximatePersistenceSize(allSegments, writeSegments));
+      return approximateTotalSize(persistenceSize, allSegments);
+   }
+
+   private CompletionStage<Long> approximatePersistenceSize(IntSet privateSegments, IntSet sharedSegments) {
+      return persistenceManager.running().approximateSize(AccessMode.PRIVATE, privateSegments)
+                               .thenCompose(privateSize -> {
+                                  if (privateSize >= 0)
+                                     return CompletableFuture.completedFuture(privateSize);
+
+                                  return persistenceManager.running().approximateSize(AccessMode.SHARED, sharedSegments);
+                               });
+   }
+
+   private long approximateTotalSize(long persistenceSize, IntSet segments) {
+      long inMemorySize = dataContainer.sizeIncludingExpired(segments);
```

Comment:
Actually `OffHeapDataContainer` and `SegmentedBoundedOffHeapDataContainer` are O(1) (since the number of segments is constant).

For local caches with eviction, `BoundedOffHeapDataContainer` does iterate over the entries, so I'll implement it in O(1).

Reply:
We could always optimize the BoundedOffHeapDataContainer by checking if the segments provided = the total segment count in the config and just return the size of the map.

Reply:
> We could always optimize the BoundedOffHeapDataContainer by checking if the segments provided = the total segment count in the config and just return the size of the map.

@wburns I tried to do this, but then I wanted to do the same check for all BoundedOffHeapDataContainer methods that receive a `segments` parameter and I noticed that the assertion was failing in some tests.

When there is a segmented store, computing the size involves calling publishEntries(segment) once for each segment, even though the cache is local, so I created https://issues.redhat.com/browse/ISPN-13574 to remove that special case and always use a single segment.